core/arch/arm/mm/core_mmu.c

```diff
@@ -988,6 +994,97 @@ static void set_pg_region(struct core_mmu_table_info *dir_info,
 	}
 }
 
+TEE_Result core_mmu_map_pages(vaddr_t vstart, paddr_t *pages, size_t num_pages,
+			      enum teecore_memtypes memtype)
+{
+	TEE_Result ret;
+	struct core_mmu_table_info tbl_info;
+	struct tee_mmap_region *mm;
+	unsigned int idx;
+	uint32_t old_attr;
+	vaddr_t vaddr = vstart;
+	size_t i;
+
+	assert(!(core_mmu_type_to_attr(memtype) & TEE_MATTR_PX));
+
+	if (vaddr & SMALL_PAGE_MASK)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	mm = find_map_by_va((void *)vaddr);
+	if (!mm || !va_is_in_map(mm, vaddr + num_pages * SMALL_PAGE_SIZE - 1))
+		panic("VA does not belong to any known mm region");
+
+	if (!core_mmu_is_dynamic_vaspace(mm))
+		panic("Trying to map into static region");
+
+	for (i = 0; i < num_pages; i++) {
+		if (pages[i] & SMALL_PAGE_MASK) {
+			ret = TEE_ERROR_BAD_PARAMETERS;
+			goto err;
+		}
+
+		while (true) {
+			if (!core_mmu_find_table(vaddr, UINT_MAX, &tbl_info)) {
+				ret = TEE_ERROR_GENERIC;
+				goto err;
+			}
+
+			idx = core_mmu_va2idx(&tbl_info, vaddr);
+			if (tbl_info.shift != SMALL_PAGE_SHIFT) {
+				if (!core_mmu_divide_block(&tbl_info, idx)) {
+					EMSG("Could not divide block into smaller tables");
+					ret = TEE_ERROR_OUT_OF_MEMORY;
+					goto err;
+				}
+			} else {
+				break;
+			}
+		}
+
+		core_mmu_get_entry(&tbl_info, idx, NULL, &old_attr);
+		if (old_attr)
+			panic("Page is already mapped");
+
+		core_mmu_set_entry(&tbl_info, idx, pages[i],
+				   core_mmu_type_to_attr(memtype));
+		vaddr += SMALL_PAGE_SIZE;
+	}
+
+	return TEE_SUCCESS;
+err:
+	if (ret != TEE_SUCCESS && i)
+		core_mmu_unmap_pages(vstart, i - 1);
```

Comment:
> > Shouldn't it be i instead of i - 1 ?
>
> Why? We call core_mmu_set_entry() at the very end of the loop. There are no goto out after that.

If you exit at 2nd round on error,  i==1 and 1 page was mapped and should be unmapped from `vstart`.
Sorry to ask again, i'm maybe wrong as you feel confident, but I don't catch it.